// @perfectmemory/ngx-contextmenu/esm2022/lib/directives/context-menu/context-menu.directive.mjs
import { Directive, HostBinding, HostListener, Input, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../services/context-menu-event/context-menu-event.service";
import * as i2 from "../../services/context-menu-stack/context-menu-stack.service";
export class ContextMenuDirective {
    constructor(contextMenuEventService, elementRef, contextMenuStackService) {
        this.contextMenuEventService = contextMenuEventService;
        this.elementRef = elementRef;
        this.contextMenuStackService = contextMenuStackService;
        /**
         * The directive must have a tabindex for being accessible
         */
        this.tabindex = '0';
        /**
         * Accessibility
         *
         * @internal
         */
        this.ariaHasPopup = 'true';
    }
    /**
     * Programmatically open the context menu
     */
    open(event) {
        if (!this.canOpen()) {
            return;
        }
        if (event instanceof MouseEvent) {
            this.onContextMenu(event);
            return;
        }
        const { x, y, height } = this.elementRef.nativeElement.getBoundingClientRect();
        this.contextMenuEventService.show({
            anchoredTo: 'position',
            x,
            y: y + height,
            contextMenu: this.contextMenu,
            value: this.contextMenuValue,
        });
    }
    /**
     * Programmatically close the context menu
     */
    close() {
        this.contextMenuStackService.closeAll();
    }
    /**
     * @internal
     */
    onContextMenu(event) {
        if (!this.canOpen()) {
            return;
        }
        this.contextMenuEventService.show({
            anchoredTo: 'position',
            contextMenu: this.contextMenu,
            x: event.clientX,
            y: event.clientY,
            value: this.contextMenuValue,
        });
        event.preventDefault();
        event.stopPropagation();
    }
    canOpen() {
        return this.contextMenu && !this.contextMenu.disabled;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuDirective, deps: [{ token: i1.ContextMenuEventService }, { token: i0.ElementRef }, { token: i2.ContextMenuStackService }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.0", type: ContextMenuDirective, selector: "[contextMenu]", inputs: { contextMenuValue: "contextMenuValue", contextMenu: "contextMenu", tabindex: "tabindex" }, host: { listeners: { "contextmenu": "onContextMenu($event)" }, properties: { "attr.tabindex": "this.tabindex", "attr.aria-haspopup": "this.ariaHasPopup" } }, exportAs: ["ngxContextMenu"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[contextMenu]',
                    exportAs: 'ngxContextMenu',
                }]
        }], ctorParameters: function () { return [{ type: i1.ContextMenuEventService }, { type: i0.ElementRef }, { type: i2.ContextMenuStackService }]; }, propDecorators: { contextMenuValue: [{
                type: Input
            }], contextMenu: [{
                type: Input
            }], tabindex: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.tabindex']
            }], ariaHasPopup: [{
                type: HostBinding,
                args: ['attr.aria-haspopup']
            }], onContextMenu: [{
                type: HostListener,
                args: ['contextmenu', ['$event']]
            }] } });
