// @perfectmemory/ngx-contextmenu/esm2022/lib/ngx-contextmenu.module.mjs
import { FullscreenOverlayContainer, OverlayContainer, OverlayModule, } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ContextMenuContentComponent } from './components/context-menu-content/context-menu-content.component';
import { ContextMenuComponent } from './components/context-menu/context-menu.component';
import { ContextMenuItemDirective } from './directives/context-menu-item/context-menu-item.directive';
import { ContextMenuDirective } from './directives/context-menu/context-menu.directive';
import * as i0 from "@angular/core";
export class ContextMenuModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuModule, declarations: [ContextMenuDirective,
            ContextMenuComponent,
            ContextMenuContentComponent,
            ContextMenuItemDirective], imports: [CommonModule, OverlayModule], exports: [ContextMenuDirective,
            ContextMenuComponent,
            ContextMenuItemDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuModule, providers: [
            { provide: OverlayContainer, useClass: FullscreenOverlayContainer },
        ], imports: [CommonModule, OverlayModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ContextMenuDirective,
                        ContextMenuComponent,
                        ContextMenuContentComponent,
                        ContextMenuItemDirective,
                    ],
                    providers: [
                        { provide: OverlayContainer, useClass: FullscreenOverlayContainer },
                    ],
                    exports: [
                        ContextMenuDirective,
                        ContextMenuComponent,
                        ContextMenuItemDirective,
                    ],
                    imports: [CommonModule, OverlayModule],
                }]
        }] });
