// @perfectmemory/ngx-contextmenu/esm2022/lib/services/context-menu-stack/context-menu-stack.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ContextMenuStackService {
    constructor() {
        this.stack = [];
    }
    /**
     * Add an item to the stack
     */
    push(value) {
        this.stack.push(value);
    }
    /**
     * Return the stack size
     */
    size() {
        return this.stack.length;
    }
    /**
     * Return true if the stack is empty
     */
    isEmpty() {
        return this.size() === 0;
    }
    /**
     * Clear the whole stack
     */
    closeAll() {
        this.stack.forEach((item) => this.dispose(item));
        this.stack = [];
    }
    /**
     * Detach and dispose sub menu's overlays of the given ContextMenuContentComponent
     */
    destroySubMenus(contextMenu) {
        if (!contextMenu) {
            return;
        }
        const index = this.stack.findIndex(({ overlayRef }) => overlayRef === contextMenu.overlayRef);
        this.stack.slice(index + 1).forEach((item) => {
            this.dispose(item);
        });
    }
    /**
     * Destroy leaf menu and return true if the destroyed menu is the root
     */
    closeLeafMenu(excludeRootMenu) {
        const item = this.disposeLastDetached();
        if (!item) {
            return false;
        }
        const moreThanOneItem = this.size() > 1;
        const isNotEmptyAndDoesNotExcludeRootMenu = !excludeRootMenu && !this.isEmpty();
        if (moreThanOneItem || isNotEmptyAndDoesNotExcludeRootMenu) {
            this.dispose(item);
        }
        if (isNotEmptyAndDoesNotExcludeRootMenu) {
            return true;
        }
        const newValue = this.disposeLastDetached();
        if (newValue) {
            newValue.contextMenuContentComponent.isLeaf = true;
        }
        return false;
    }
    /**
     * Starting by most recent items, dispose all detached item and return the most recent attached one
     */
    disposeLastDetached() {
        let item = this.last();
        if (!item) {
            return;
        }
        while (item &&
            this.size() > 1 &&
            item.overlayRef &&
            this.isDetached(item)) {
            this.dispose(item);
            this.pop();
            item = this.last();
        }
        return item;
    }
    last() {
        if (this.isEmpty()) {
            return;
        }
        return this.stack[this.stack.length - 1];
    }
    pop() {
        const value = this.stack.pop();
        return this.dispose(value);
    }
    isDetached(item) {
        return !item.overlayRef.hasAttached();
    }
    dispose(item) {
        if (item) {
            item.overlayRef.detach();
            item.overlayRef.dispose();
        }
        return item;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuStackService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuStackService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuStackService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
