// @perfectmemory/ngx-contextmenu/esm2022/lib/components/context-menu-content/context-menu-content.component.mjs
import { ActiveDescendantKeyManager } from '@angular/cdk/a11y';
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, HostListener, Input, Output, QueryList, ViewChildren, } from '@angular/core';
import { Subscription } from 'rxjs';
import { evaluateIfFunction } from '../../helper/evaluate';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const ARROW_LEFT_KEYCODE = 37;
const ARROW_RIGHT_KEYCODE = 39;
/**
 * For testing purpose only
 */
export const TESTING_WRAPPER = {
    ActiveDescendantKeyManager,
};
export class ContextMenuContentComponent {
    // TODO: should be private but issue in spec with NullInjectorError: No provider for ElementRef!
    constructor(_elementRef) {
        this._elementRef = _elementRef;
        /**
         * The list of `ContextMenuItemDirective` that represent each menu items
         */
        this.menuDirectives = [];
        /**
         * A CSS class to apply a theme to the the menu
         */
        this.menuClass = '';
        /**
         * Wether the instance is a leaf menu or not
         */
        this.isLeaf = false;
        /**
         * Emit when a menu item is selected
         */
        this.execute = new EventEmitter();
        /**
         * Emit when a sub menu is opened
         */
        this.openSubMenu = new EventEmitter();
        /**
         * Emit when a leaf menu is closed
         */
        this.closeLeafMenu = new EventEmitter();
        /**
         * Emit when sub menus are closed
         */
        this.closeSubMenus = new EventEmitter();
        /**
         * Emit when all menus is closed
         */
        this.closeAllMenus = new EventEmitter();
        this.subscription = new Subscription();
    }
    /**
     * @internal
     */
    ngOnInit() {
        this.setupDirectives();
    }
    /**
     * @internal
     */
    ngAfterViewInit() {
        this.overlayRef?.updatePosition();
    }
    /**
     * @internal
     */
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    /**
     * @internal
     */
    onKeyArrowDownOrUp(event) {
        if (!this.isLeaf) {
            return;
        }
        this.keyManager.onKeydown(event);
    }
    /**
     * @internal
     */
    onKeyArrowRight(event) {
        if (!this.isLeaf) {
            return;
        }
        if (this.dir === 'rtl') {
            this.closeActiveItemSubMenu(event);
            return;
        }
        this.openActiveItemSubMenu(event);
    }
    /**
     * @internal
     */
    onKeyArrowLeft(event) {
        if (!this.isLeaf) {
            return;
        }
        if (this.dir === 'rtl') {
            this.openActiveItemSubMenu(event);
            return;
        }
        this.closeActiveItemSubMenu(event);
    }
    /**
     * @internal
     */
    onKeyEnterOrSpace(event) {
        if (!this.isLeaf) {
            return;
        }
        if (!this.keyManager.activeItem) {
            return;
        }
        this.onMenuItemSelect(this.keyManager.activeItem, event);
    }
    /**
     * @internal
     */
    onKeyArrowEscape(event) {
        if (!this.isLeaf) {
            return;
        }
        this.closeActiveItemSubMenu(event);
    }
    /**
     * @internal
     */
    onClickOrRightClick(event) {
        if (event.type === 'click' && event.button === 2) {
            return;
        }
        if (this._elementRef.nativeElement.contains(event.target)) {
            return;
        }
        this.closeAllMenus.emit({ event });
    }
    /**
     * @internal
     */
    stopEvent(event) {
        event.stopPropagation();
    }
    /**
     * @internal
     */
    isMenuItemDisabled(menuItem) {
        return evaluateIfFunction(menuItem.disabled, this.value);
    }
    /**
     * @internal
     */
    isMenuItemVisible(menuItem) {
        return evaluateIfFunction(menuItem.visible, this.value);
    }
    /**
     * @internal
     */
    onOpenSubMenu(menuItem, event) {
        if (!menuItem.subMenu) {
            this.closeSubMenus.next();
        }
        if (this.keyManager.activeItemIndex === null || !menuItem.subMenu) {
            return;
        }
        const anchorElementRef = this.liElementRefs.toArray()[this.keyManager.activeItemIndex];
        const anchorElement = anchorElementRef && anchorElementRef.nativeElement;
        if (anchorElement && event instanceof KeyboardEvent) {
            this.openSubMenu.emit({
                anchoredTo: 'element',
                anchorElement,
                contextMenu: menuItem.subMenu,
                value: this.value,
                parentContextMenu: this,
            });
        }
        else if (event.currentTarget) {
            this.openSubMenu.emit({
                anchoredTo: 'element',
                anchorElement: event.currentTarget,
                contextMenu: menuItem.subMenu,
                value: this.value,
                parentContextMenu: this,
            });
        }
        else {
            this.openSubMenu.emit({
                anchoredTo: 'position',
                x: event.clientX,
                y: event.clientY,
                contextMenu: menuItem.subMenu,
                value: this.value,
            });
        }
    }
    /**
     * @internal
     */
    onMenuItemSelect(menuItem, event) {
        this.cancelEvent(event);
        this.onOpenSubMenu(menuItem, event);
        if (!menuItem.subMenu) {
            this.triggerExecute(menuItem, event);
        }
    }
    triggerExecute(menuItem, event) {
        menuItem.triggerExecute(event, this.value);
    }
    setupDirectives() {
        this.menuDirectives.forEach((menuDirective) => {
            menuDirective.value = this.value;
            this.subscription.add(menuDirective.execute.subscribe((event) => this.execute.emit({ ...event, menuDirective })));
        });
        const queryList = new QueryList();
        queryList.reset(this.menuDirectives);
        this.keyManager = new TESTING_WRAPPER.ActiveDescendantKeyManager(queryList).withWrap();
    }
    openActiveItemSubMenu(event) {
        if (this.keyManager.activeItemIndex === null) {
            return;
        }
        this.cancelEvent(event);
        if (this.keyManager.activeItem) {
            this.onOpenSubMenu(this.keyManager.activeItem, event);
        }
    }
    closeActiveItemSubMenu(event) {
        if (this.keyManager.activeItemIndex === null) {
            return;
        }
        this.cancelEvent(event);
        this.closeLeafMenu.emit({
            excludeRootMenu: this.dir === 'rtl'
                ? event.keyCode === ARROW_RIGHT_KEYCODE
                : event.keyCode === ARROW_LEFT_KEYCODE,
            event,
        });
    }
    cancelEvent(event) {
        if (!event || !event.target) {
            return;
        }
        const target = event.target;
        if (['INPUT', 'TEXTAREA', 'SELECT'].includes(target.tagName) ||
            target.isContentEditable) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuContentComponent, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.0.0", type: ContextMenuContentComponent, selector: "context-menu-content", inputs: { menuDirectives: "menuDirectives", value: "value", dir: "dir", parentContextMenu: "parentContextMenu", menuClass: "menuClass", overlayRef: "overlayRef", isLeaf: "isLeaf" }, outputs: { execute: "execute", openSubMenu: "openSubMenu", closeLeafMenu: "closeLeafMenu", closeSubMenus: "closeSubMenus", closeAllMenus: "closeAllMenus" }, host: { listeners: { "window:keydown.ArrowDown": "onKeyArrowDownOrUp($event)", "window:keydown.ArrowUp": "onKeyArrowDownOrUp($event)", "window:keydown.ArrowRight": "onKeyArrowRight($event)", "window:keydown.ArrowLeft": "onKeyArrowLeft($event)", "window:keydown.Enter": "onKeyEnterOrSpace($event)", "window:keydown.Space": "onKeyEnterOrSpace($event)", "window:keydown.Escape": "onKeyArrowEscape($event)", "document:click": "onClickOrRightClick($event)", "document:contextmenu": "onClickOrRightClick($event)" }, properties: { "attr.dir": "this.dir" } }, viewQueries: [{ propertyName: "liElementRefs", predicate: ["li"], descendants: true }], ngImport: i0, template: "<div class=\"ngx-contextmenu\" [ngClass]=\"menuClass\" [attr.dir]=\"dir\">\n  <ul #menu class=\"ngx-contextmenu--dropdown-menu\" role=\"menu\" tabindex=\"0\">\n    <li\n      #li\n      *ngFor=\"let menuDirective of menuDirectives; let i = index\"\n      [class.disabled]=\"isMenuItemDisabled(menuDirective)\"\n      [class.divider]=\"menuDirective.divider\"\n      [class.active]=\"\n        menuDirective.isActive && !isMenuItemDisabled(menuDirective)\n      \"\n      [attr.role]=\"menuDirective.divider ? 'separator' : undefined\"\n      role=\"none\"\n    >\n      <hr *ngIf=\"menuDirective.divider\" role=\"separator\" />\n      <button\n        *ngIf=\"!menuDirective.divider && !menuDirective.passive\"\n        type=\"button\"\n        role=\"menuitem\"\n        class=\"ngx-contextmenu--item-content\"\n        [attr.aria-haspopup]=\"!!menuDirective.subMenu\"\n        [class.ngx-contextmenu--parent-menu]=\"!!menuDirective.subMenu\"\n        [class.active]=\"\n          menuDirective.isActive && !isMenuItemDisabled(menuDirective)\n        \"\n        [class.disabled]=\"isMenuItemDisabled(menuDirective)\"\n        [attr.disabled]=\"isMenuItemDisabled(menuDirective) ? 'disabled' : null\"\n        (click)=\"onMenuItemSelect(menuDirective, $event)\"\n        (mouseenter)=\"onOpenSubMenu(menuDirective, $event)\"\n      >\n        <ng-template\n          [ngTemplateOutlet]=\"menuDirective.template\"\n          [ngTemplateOutletContext]=\"{ item: value }\"\n        ></ng-template>\n      </button>\n\n      <span\n        *ngIf=\"!menuDirective.divider && menuDirective.passive\"\n        class=\"ngx-contextmenu--item-content ngx-contextmenu--item-content-passive\"\n        role=\"menuitem\"\n        (click)=\"stopEvent($event)\"\n        (contextmenu)=\"stopEvent($event)\"\n        [class.disabled]=\"isMenuItemDisabled(menuDirective)\"\n      >\n        <ng-template\n          [ngTemplateOutlet]=\"menuDirective.template\"\n          [ngTemplateOutletContext]=\"{ item: value }\"\n        ></ng-template>\n      </span>\n    </li>\n  </ul>\n</div>\n", dependencies: [{ kind: "directive", type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i1.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuContentComponent, decorators: [{
            type: Component,
            args: [{ selector: 'context-menu-content', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"ngx-contextmenu\" [ngClass]=\"menuClass\" [attr.dir]=\"dir\">\n  <ul #menu class=\"ngx-contextmenu--dropdown-menu\" role=\"menu\" tabindex=\"0\">\n    <li\n      #li\n      *ngFor=\"let menuDirective of menuDirectives; let i = index\"\n      [class.disabled]=\"isMenuItemDisabled(menuDirective)\"\n      [class.divider]=\"menuDirective.divider\"\n      [class.active]=\"\n        menuDirective.isActive && !isMenuItemDisabled(menuDirective)\n      \"\n      [attr.role]=\"menuDirective.divider ? 'separator' : undefined\"\n      role=\"none\"\n    >\n      <hr *ngIf=\"menuDirective.divider\" role=\"separator\" />\n      <button\n        *ngIf=\"!menuDirective.divider && !menuDirective.passive\"\n        type=\"button\"\n        role=\"menuitem\"\n        class=\"ngx-contextmenu--item-content\"\n        [attr.aria-haspopup]=\"!!menuDirective.subMenu\"\n        [class.ngx-contextmenu--parent-menu]=\"!!menuDirective.subMenu\"\n        [class.active]=\"\n          menuDirective.isActive && !isMenuItemDisabled(menuDirective)\n        \"\n        [class.disabled]=\"isMenuItemDisabled(menuDirective)\"\n        [attr.disabled]=\"isMenuItemDisabled(menuDirective) ? 'disabled' : null\"\n        (click)=\"onMenuItemSelect(menuDirective, $event)\"\n        (mouseenter)=\"onOpenSubMenu(menuDirective, $event)\"\n      >\n        <ng-template\n          [ngTemplateOutlet]=\"menuDirective.template\"\n          [ngTemplateOutletContext]=\"{ item: value }\"\n        ></ng-template>\n      </button>\n\n      <span\n        *ngIf=\"!menuDirective.divider && menuDirective.passive\"\n        class=\"ngx-contextmenu--item-content ngx-contextmenu--item-content-passive\"\n        role=\"menuitem\"\n        (click)=\"stopEvent($event)\"\n        (contextmenu)=\"stopEvent($event)\"\n        [class.disabled]=\"isMenuItemDisabled(menuDirective)\"\n      >\n        <ng-template\n          [ngTemplateOutlet]=\"menuDirective.template\"\n          [ngTemplateOutletContext]=\"{ item: value }\"\n        ></ng-template>\n      </span>\n    </li>\n  </ul>\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { menuDirectives: [{
                type: Input
            }], value: [{
                type: Input
            }], dir: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.dir']
            }], parentContextMenu: [{
                type: Input
            }], menuClass: [{
                type: Input
            }], overlayRef: [{
                type: Input
            }], isLeaf: [{
                type: Input
            }], execute: [{
                type: Output
            }], openSubMenu: [{
                type: Output
            }], closeLeafMenu: [{
                type: Output
            }], closeSubMenus: [{
                type: Output
            }], closeAllMenus: [{
                type: Output
            }], liElementRefs: [{
                type: ViewChildren,
                args: ['li']
            }], onKeyArrowDownOrUp: [{
                type: HostListener,
                args: ['window:keydown.ArrowDown', ['$event']]
            }, {
                type: HostListener,
                args: ['window:keydown.ArrowUp', ['$event']]
            }], onKeyArrowRight: [{
                type: HostListener,
                args: ['window:keydown.ArrowRight', ['$event']]
            }], onKeyArrowLeft: [{
                type: HostListener,
                args: ['window:keydown.ArrowLeft', ['$event']]
            }], onKeyEnterOrSpace: [{
                type: HostListener,
                args: ['window:keydown.Enter', ['$event']]
            }, {
                type: HostListener,
                args: ['window:keydown.Space', ['$event']]
            }], onKeyArrowEscape: [{
                type: HostListener,
                args: ['window:keydown.Escape', ['$event']]
            }], onClickOrRightClick: [{
                type: HostListener,
                args: ['document:click', ['$event']]
            }, {
                type: HostListener,
                args: ['document:contextmenu', ['$event']]
            }] } });
