// @perfectmemory/ngx-contextmenu/esm2022/lib/services/context-menu/context-menu.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../context-menu-event/context-menu-event.service";
import * as i2 from "../context-menu-stack/context-menu-stack.service";
/**
 * Programmatically open a ContextMenuComponent to a X/Y position
 */
export class ContextMenuService {
    constructor(contextMenuEventService, contextMenuStackService) {
        this.contextMenuEventService = contextMenuEventService;
        this.contextMenuStackService = contextMenuStackService;
    }
    /**
     * Show the given `ContextMenuComponent` at a specified X/Y position
     */
    show(contextMenu, options = { x: 0, y: 0 }) {
        this.contextMenuEventService.show({
            anchoredTo: 'position',
            contextMenu,
            value: options.value,
            x: options.x,
            y: options.y,
        });
    }
    /**
     * Close all open `ContextMenuComponent`
     */
    closeAll() {
        this.contextMenuStackService.closeAll();
    }
    /**
     * Return true if any `ContextMenuComponent` is open
     */
    hasOpenMenu() {
        return !this.contextMenuStackService.isEmpty();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuService, deps: [{ token: i1.ContextMenuEventService }, { token: i2.ContextMenuStackService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: ContextMenuService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.ContextMenuEventService }, { type: i2.ContextMenuStackService }]; } });
